### PR TITLE
buildBazelPackage: autodetect nix toolchain instead of Xcode one on macOS

### LINKED ISSUE #55991: bazel-watcher 0.9.0 broken in macOS
## Issue description

By looking at: https://hydra.nixos.org/job/nixpkgs/trunk/bazel-watcher.x86_64-darwin, it seems `bazel-watcher` has never worked on `macOS`. It seems as if the derivation cannot find the standard header files when building. @uri-canva @kalbasit any ideas how to get this working? 

### Steps to reproduce

```
❯ nix-build -E "import <nixpkgs> {}" -A pkgs.bazel-watcher
```

```
In file included from external/com_google_protobuf/src/google/protobuf/stubs/stringpiece.cc:30:
external/com_google_protobuf/src/google/protobuf/stubs/stringpiece.h:147:10: fatal error: 'iosfwd' file not found
#include <iosfwd>
         ^~~~~~~~
1 error generated.
Target //ibazel:ibazel failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 12.723s, Critical Path: 1.39s
INFO: 8 processes: 8 processwrapper-sandbox.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
builder for '/nix/store/8dg74gn93va1s624akkfyfmyinwsmj3i-bazel-watcher-0.9.0.drv' failed with exit code 1
```
## Technical details

```
 - system: `"x86_64-darwin"`
 - host os: `Darwin 17.7.0, macOS 10.13.6`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.1.3`
 - channels(danielgarzon): `"nixos-19.03pre169108.36f31600749, nixpkgs-19.03pre169106.1a88aa9e0cd"`
 - nixpkgs: `/Users/danielgarzon/.nix-defexpr/channels/nixpkgs`
```

### PR DESCRIPTION
**Pull request by uri-canva**
###### Motivation for this change

Certain derivations fail to build with sandboxing enabled, depending on how Xcode / CLT are configured and installed. See #55991.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

fixes #55991

### CODE REVIEW

**Comment by uri-canva** [comment]
Please build `bazel-watcher`, `tensorflow` and `bazel-deps`.

**Comment by kalbasit** [comment]
@uri-canva does this work for you now for building bazel-watcher on Darwin?
@LnL7 ofborg built the previous one correctly, but it failed on Hydra. Is there a way to get this tested on hydra before we merge?

@GrahamcOfBorg build bazel-watcher tensorflow bazel-deps

**Comment by uri-canva** [comment]
`bazel-watcher` built correctly for me on macOS before, assuming Xcode was installed and set up correctly. I broke my Xcode installation to repro the issue where headers couldn't be found, then made `bazel-watcher` build correctly without requiring Xcode.

File: pkgs/build-support/build-bazel-package/default.nix
Comment (kalbasit) [inline pkgs/build-support/build-bazel-package/default.nix]: @uri-canva why are all of these passed to bazel? Are they necessary? Can you perhaps add a comment (in the code) describing what are these doing and why are they necessary? Thanks!

**Review by kalbasit — CHANGES_REQUESTED** [review]

**Comment by kalbasit** [comment]
looks like bazel-deps has failed. Can you try it locally if it fails for you for this PR? What about on master?

I would try it myself, but I'm currently at a hotel with a very bad connection :(

```
ERROR: /build/output/external/bazel_tools/tools/jdk/BUILD:188:1: SkylarkAction external/bazel_tools/tools/jdk/platformclasspath_classes/DumpPlatformClassPath.class failed (Exit 1) javac failed: error executing command external/remotejdk_linux/bin/javac -source 8 -target 8 -Xlint:-options -cp external/remotejdk_linux/lib/tools.jar -d bazel-out/host/bin/external/bazel_tools/tools/jdk/platformclasspath_classes ... (remaining 1 argument(s) skipped)
Use --sandbox_debug to see verbose messages from the sandbox
src/main/tools/linux-sandbox-pid1.cc:427: "execvp(external/remotejdk_linux/bin/javac, 0x4f5c10)": No such file or directory
Target //src/scala/com/github/johnynek/bazel_deps:parseproject_deploy.jar failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 24.515s, Critical Path: 0.09s
INFO: 1 process: 1 linux-sandbox.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
builder for '/nix/store/dnjxdqnm32yzjsamxmmhz1yy3qszib9m-bazel-deps-2019-02-01.drv' failed with exit code 1
error: build of '/nix/store/dnjxdqnm32yzjsamxmmhz1yy3qszib9m-bazel-deps-2019-02-01.drv' failed
```

**Comment by uri-canva** [comment]
Yeah I noticed, that's broken by the Bazel update on master: https://hydra.nixos.org/jobset/nixpkgs/trunk#tabs-jobs, we can't test this PR on that.

**Comment by uri-canva** [comment]
Oh sorry you're right, the failure is actually different, I will have a look.

**Comment by uri-canva** [comment]
Ok, tested it locally, the failure I get is the same on this branch as it is on master, so I think it's ok. We can't verify on hydra because it doesn't have a log.

**Comment by kalbasit** [comment]
@uri-canva Fair enough, we'll let Hydra test it out once we merge it down. Can you please address my comment I left inline and I'll wait for ofborg to test it out on Darwin before I merge.

**Comment by uri-canva** [comment]
Added comments.

**Review by kalbasit — APPROVED** [review]
LGTM

**Comment by kalbasit** [comment]
@GrahamcOfBorg build bazel-watcher

File: pkgs/build-support/build-bazel-package/default.nix
Comment (Profpatsch) [inline pkgs/build-support/build-bazel-package/default.nix]: Does this work on Linux?

**Comment by zimbatm** [comment]
`nix-review pr 56033` produces the following output:
```
builder for '/nix/store/c8n9njq6jmdiy4abjwz1rxizfidhmg9x-bazel-deps-2019-02-01.drv' failed with exit code 1; last 10 log lines:
  ERROR: /build/output/external/bazel_tools/tools/jdk/BUILD:188:1: SkylarkAction external/bazel_tools/tools/jdk/platformclasspath_classes/DumpPlatformClassPath.class failed (Exit 1) javac failed: error executing command external/remotejdk_linux/bin/javac -source 8 -target 8 -Xlint:-options -cp external/remotejdk_linux/lib/tools.jar -d bazel-out/host/bin/external/bazel_tools/tools/jdk/platformclasspath_classes ... (remaining 1 argument(s) skipped)
  
  Use --sandbox_debug to see verbose messages from the sandbox
  src/main/tools/linux-sandbox-pid1.cc:427: "execvp(external/remotejdk_linux/bin/javac, 0x4f5c10)": No such file or directory
  Target //src/scala/com/github/johnynek/bazel_deps:parseproject_deploy.jar failed to build
  Use --verbose_failures to see the command lines of failed build steps.
  INFO: Elapsed time: 18.244s, Critical Path: 0.41s
  INFO: 7 processes: 7 linux-sandbox.
  FAILED: Build did NOT complete successfully
  FAILED: Build did NOT complete successfully
cannot build derivation '/nix/store/0si8yncz9l72d2dja62ryf451pdcb4wk-env.drv': 1 dependencies couldn't be built
```

**Review by ghost — APPROVED** [review]
Given that bazel-deps is both broken for Linux on master and 18.09, I think it's safe to merge this as an incremental improvement. If anyone wants a working Linux build they will have to do it themselves.

File: pkgs/build-support/build-bazel-package/default.nix
Comment (kalbasit) [reply pkgs/build-support/build-bazel-package/default.nix]: This is a no-op on Linux. See it in the source code [here](https://github.com/bazelbuild/bazel/blob/71bc38fe2d165fe523b19d7879096d3fdd2086fc/tools/cpp/cc_configure.bzl#L53-L55).

**Comment by abbradar** [comment]
@uri-canva While meddling with Tensorflow I noticed that the code that adds flags from environment variables isn't really needed - for example `NIX_LDFLAGS` just works (and is isn't even handled there). Maybe it makes a difference for different derivations? What did it fix for you?

**Comment by abbradar** [comment]
Nevermind, I found out this is because Nix hacks are disabled for some packages. Instead of this I propose:

1. Drop one part of Nix hacks (disabling timestamp checking) because somehow it just works without it for Tensorflow now, and I assume for other packages too;
2. Always use Nix hacks for building Nix packages because that's much easier w.r.t various environment variables that Nix uses.

**Comment by uri-canva** [comment]
The nix hacks disable the environment sanitisation logic in Bazel, which means Bazel will reuse cached results regardless of whether they used environment variables that changed in the meantime or not. That should be ok as long as we never use Bazel to cache outputs within nix, which is a reasonable assumption since we would be using the nix store to cache outputs anyway.

**Pull request merged**
